src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -474,58 +473,69 @@ private static bool IsValidIPv6Host(string hostText)
                     // [::1] is the shortest valid IPv6 host
                     if (i < 4)
                     {
-                        return false;
+                        BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
+                    }
+                    else if (i + 1 < hostText.Length)
+                    {
+                        // Tail call
+                        ValidateHostPort(hostText, i + 1);
                     }
-                    return IsValidHostPort(hostText, i + 1);
+                    return;
                 }
 
                 if (!IsHex(ch) && ch != ':' && ch != '.')
                 {
-                    return false;
+                    BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
                 }
             }
 
             // Must contain a ']'
-            return false;
+            BadHttpRequestException.Throw(RequestRejectionReason.InvalidHostHeader, hostText);
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static bool IsValidHostPort(string hostText, int offset)
+        private static void ValidateHostPort(string hostText, int offset)
         {
-            if (offset == hostText.Length)
-            {
-                return true;
-            }
-
-            if (hostText[offset] != ':' || hostText.Length == offset + 1)
+            var firstChar = hostText[offset];
+            offset++;
+            if (firstChar != ':' || (uint)offset >= (uint)hostText.Length)
```

Comment:
The other uint casts look to be eliminating a comparison which I'm OK with. This looks like it's just for the JIT though.

Reply:
Sorta, its verifying `offset` is positive so the `var i = offset` in the `for` loop so the bounds checks can be eliminated in the for loop below.

However, the Jit still doesn't eliminate the bounds check at this time; so I can remove it? (doesn't really like doing it for things that start with variables)

Reply:
Removed